File: C/39-combination-sum/main.go
package main

import (
	"fmt"
)

func combinationSum(candidates []int, target int) [][]int {
	buf := map[int][][]int{}
	buf[target] = [][]int{{target}}

	for i := target; i >= 0; i-- {
		v, ok := buf[i]

		if ok {
			for _, c := range candidates {
				b := make([][]int, len(v)+len(buf[i-c]))
				idx := 0
				for _, ll := range buf[i-c] {
					b[idx] = ll
					idx++
				}
				for _, vv := range v {
					if c <= vv[0] {
						t := make([]int, len(vv)+1)
						copy(t[1:], vv)
						t[0] = c
						b[idx] = t
						idx++
					}
				}
				buf[i-c] = b[0:idx]
			}
		}
	}

	for i := 0; i < len(buf[0]); i++ {
		buf[0][i] = buf[0][i][0 : len(buf[0][i])-1]
	}
	return buf[0]
}

func combinationSumV2(candidates []int, target int) [][]int {
	buf := map[int][][]int{}
	buf[0] = [][]int{{0}}

	for i := 0; i <= target; i++ {
		v, ok := buf[i]

		if ok {
			for _, c := range candidates {
				b := make([][]int, len(v)+len(buf[i+c]))
				idx := 0
				for _, ll := range buf[i+c] {
					b[idx] = ll
					idx++
				}
				for _, vv := range v {
					if c >= vv[len(vv)-1] {
						t := make([]int, len(vv)+1)
						copy(t, vv)
						t[len(vv)] = c
						b[idx] = t
						idx++
					}
				}
				buf[i+c] = b[0:idx]
			}
		}
	}

	for i := 0; i < len(buf[target]); i++ {
		buf[target][i] = buf[target][i][1:]
	}
	return buf[target]
}

func main() {
	fmt.Println(combinationSum([]int{2, 3, 5}, 8))
}

File: C/39-combination-sum/main.rs
struct Solution {}

impl Solution {
  pub fn combination_sum(candidates: Vec<i32>, target: i32) -> Vec<Vec<i32>> {
    let mut candidates = candidates;
    candidates.sort();
    let mut dp: Vec<Vec<Vec<Vec<i32>>>> =
      vec![vec![Vec::new(); candidates.len()]; target as usize + 1];

    candidates.iter().enumerate().for_each(|(i, &v)| {
      if v <= target {
        dp[v as usize][i].push(vec![v]);
      }
    });
    (0..target as usize).for_each(|base| {
      let mut idx: usize = 0;
      while idx < candidates.len() {
        if dp[base][idx].len() == 0 {
          idx += 1;
          continue;
        }

        let mut idx2: usize = idx;

        while idx2 < candidates.len() {
          if base as i32 + candidates[idx2] > target {
            break;
          }

          let mut idx3: usize = 0;
          while idx3 < dp[base][idx].len() {
            let mut vv = dp[base][idx][idx3].clone();
            vv.push(candidates[idx2]);
            dp[base + candidates[idx2] as usize][idx2].push(vv);
            idx3 += 1;
          }

          idx2 += 1;
        }
        idx += 1;
      }
    });

    let mut ret: Vec<Vec<i32>> = Vec::new();
    dp[target as usize].iter().for_each(|vvv| {
      vvv.iter().for_each(|vv| {
        ret.push(vv.clone());
      });
    });
    ret
  }
}

fn main() {
  println!("{:?}", Solution::combination_sum(vec![2, 3, 6, 7], 7));
}
